docs/book/how-do-i-test-policies.md

```diff
@@ -204,6 +204,117 @@ $ opa test --format=json pass_fail_error_test.rego
 ]
 ```
 
+## Data Mocking
+
+OPA's `with` keyword can be used to mock the data document. Both base and virtual documents can be mocked. Below is a simple policy that depends on the data document.
+
+**authz.rego**:
+
+```ruby
+package authz
+
+allow {
+    x := data.policies[_]
+    x.name = "test_policy"
+    matches_role(input.role)
+}
+
+matches_role(my_role) {
+    data.roles[my_role][_] = input.user
+}
+```
+
+Below is the Rego file to test the above policy.
+
+**authz_test.rego**:
+
+```ruby
+package authz
+
+policies = [{"name": "test_policy"}]
+roles = {"admin": ["alice"]}
+
+test_allow_with_data {
+    allow  with input as {"user": "alice", "role": "admin"}  with data.policies as policies  with data.roles as roles
+}
+```
+
+To exercise the policy, run the `opa test` command.
+
+```bash
+$ opa test -v authz.rego authz_test.rego
+data.authz.test_allow_with_data: PASS (697ns)
+--------------------------------------------------------------------------------
+PASS: 1/1
+```
+
+Below is an example to mock a rule.
+
+**authz.rego**:
+
+```ruby
+package authz
+
+allow1 {
+    allow2
+}
+
+allow2 {
+    2 = 1
+}
+```
+
+**authz_test.rego**:
+
+```ruby
+package authz
+
+test_mock_rule {
+    allow1 with allow2 as true
+}
+```
+
+```bash
+$  opa test -v authz.rego authz_test.rego
+data.authz.test_mock_rule: PASS (328ns)
+--------------------------------------------------------------------------------
+PASS: 1/1
+```
+
+Functions with arguments cannot be mocked. For example, in the below policy the function `cannot_mock` cannot be mocked.
+
+**authz.rego**:
+
+```ruby
+package authz
+
+invalid_mock {
+    cannot_mock(input.label)
+}
+
+cannot_mock(label) {
+    label = "test_label"
+}
+```
+
+**authz_test.rego**:
+
+```ruby
+package authz
+
+test_invalid_mock {
+    invalid_mock with input as {"label": "test_label"} with cannot_mock as true
+}
+```
+
+```bash
+$  opa test -v authz.rego authz_test.rego
+1 error occurred: authz.rego:7: rego_compile_error: with data target cannot be used to mock a function with args
```

Comment:
This will need to be updated if we change the error message as mentioned in my comment above. Also there's an extra space after the `$`.

Reply:
Updated doc